PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/FullGfaReader.java

```diff
@@ -3,163 +3,229 @@
+   * Parse the header part of the GFA file, which contains all of the genomes.
+   *
+   * @param chars the characters of the header.
    */
-  private void handleL(Scanner sc) {
-    int parentId = sc.nextInt();
-    sc.next();
-    int childId = sc.nextInt();
-    Node parent = originalGraph.getNode(parentId);
-    Node child = originalGraph.getNode(childId);
-    parent.addOutlink(child.getId());
-    child.addInlink(parent.getId());
-    sc.nextLine();
+  private void parseHeader(char[] chars) {
+    int index = 0;
+    index = skipTillCharacter(chars, index, ':', 2) + 1;
+    int start = index;
+    while (index < chars.length) {
+      index = skipTillCharacter(chars, index, ';', 1);
+      genomes.add(new String(chars, start, index - start));
+      start = index + 1;
+      index += 2;
+    }
   }
-  
+
   /**
-   * This method extracts the single genomes from the ORI tag.
-   * @param line the full ORI tag
-   * @return an arrayList of the names of the genomes.
+   * Skips all characters till the given character is found.
+   *
+   * @param chars      the array of characters.
+   * @param startIndex the index at which to start in the array.
+   * @param ch         the character to skip to.
+   * @param amount     the amount of times to skip the character.
+   * @return the index of the first occurence of the character after the given start index.
    */
-  private ArrayList<String> extractGenomes(String line) {
-    String[] words = line.split(":");
-    if (words[0].endsWith("ORI")) {
-      String[] gens = words[2].split(";");
-      ArrayList<String> gensAr = new ArrayList<>(Arrays.asList(gens));
-      return gensAr;
+  private int skipTillCharacter(char[] chars, int startIndex, char ch, int amount) {
+    int index = startIndex;
+    for (int i = 0; i < amount; i++) {
+      ++index;
+      while (chars[index] != ch) {
+        ++index;
+      }
     }
-    return null;
+    return index;
   }
-  
+
   /**
-   * This method extracts the alignment integer from the START tag.
-   * @param line the full START tag
-   * @return the alignment integer of the tag
+   * Get a node, or create it if it doesn't exist yet.
+   *
+   * @param id the id of the node.
+   * @return the node.
    */
-  private int extractStart(String line) {
-    String[] words = line.split(":");
-    if (words[0].endsWith("START")) {
-      return Integer.parseInt(words[2]);
+  private Node getNode(int id) {
+    Node node = nodes.get(id);
+    if (node == null) {
+      node = new Node(id, 1, new ArrayList<>(), 0);
+      nodes.put(id, node);
     }
-    return 0;
+    return nodes.get(id);
   }
 
   /**
    * Method which returns the read graph.
    *
-   * @return The graph.
+   * @return the graph.
    */
   public OriginalGraph getGraph() {
-    originalGraph.setGenoms(genoms);
+    if (originalGraph == null) {
+      originalGraph = new OriginalGraph(nodes, 1, genomes);
+    }
     return originalGraph;
   }
```

Comment:
Yes that could be better.
This is not a singleton pattern, it has a public constructor and doesn't store a static reference to itself.
This class has to store the hasmap of nodes because it must be appended to when I try to get an ID which doesn't exist in the hasmap yet. It is not desirable for the OriginalGraph to have this behavior.
I only store the graph in a variable because like this it will return the same instance of the graph if you call getGraph multiple times (otherwise there could potentially be multiple original graphs with the same hashmap of nodes and list of genomes). This shouldn't really happen, but it is a bit more safe and doesn't cost anything.